### PR TITLE
DS-3850 Create new submissions from files or identifiers

### PR DESCRIPTION
**Pull request by abollini**
https://jira.duraspace.org/browse/DS-3850

Complete the WorkspaceItem endpoint
- add findByEPerson search method
- introduce pre and post functions to allow one step to react to changes done/requested to other steps/sections
- implement the pre/post functions in the Describe step to support automatic enhancement of a submission looking to the uploaded files (the grobid integration will be a separate PR) or provided metadata (identifiers)
- implement the creation from a bibliographic file (bibtex via BTE)

This PR has been extracted from #2102 and not really tested. IT for the enhancements will be added in the next days but I will appreciate early feedback and help from other in testing/review

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/submit/extraction/MetadataExtractor.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.util.List;
+
+import gr.ekt.bte.dataloader.FileDataLoader;
+import org.dspace.services.ConfigurationService;
+
+/**
+ * @author Luigi Andrea Pascarelli (luigiandrea.pascarelli at 4science.it)
+ */
+public class MetadataExtractor {
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/submit/extraction/MetadataExtractor.java]: Please add Javadocs to describe this class

**Review by tdonohue — COMMENTED** [review]
Overall, the code here is looking good.  However, I do see that the code in this PR is the first place that the usability problem recently described in [DS-4002](https://jira.duraspace.org/browse/DS-4002) begins to appear.  There are places in the PR code where metadata is automatically loaded on the server-side and saved into the WorkspaceItem without any user interaction, e.g. in [`MetadataStep.doPostProcessing()`](https://github.com/DSpace/DSpace/pull/2187/files#diff-6f6007f254fce6ba67b2c2e0d63ec78dR70)

I did add some inline comments below of places where Javadocs would be useful (essentially on every new class/interface).  Obviously, some IT tests would also be good, as there are not any yet included.

In summary, I think the code here looks fine for now. But, we need to be aware that this code may need future refactoring (perhaps after the Submission code is all merged) to resolve/fix DS-4002.

**Comment by benbosman** [comment]
My impressions of the PR are in line with Tim's review.

After a detailed review of the code in this PR, the code is looking good, but it seems that the functionality in this PR is almost entirely focussed on changes for the automatic enhancement of a submission.
The implementation strategy that was used here (this interpretation is based on the code only due to the lack of comments which were mentioned by @tdonohue ) is:
* PreProcessing functions are created to initialize the listeners for automatically adding the metadata
* PostProcessing functions are created to actually add the automatic metadata

But these PreProcessing and PostProcessing are embedded deeply in the submission backend, and currently don't allow any changes to trigger user interaction. A sample when uploading a file can be seen at https://github.com/DSpace/DSpace/blob/c0066d2a2fa4afacadba2bf36f20d94cd818217c/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/WorkspaceItemRestRepository.java#L244 where the PostProcessing is not supposed to happen prior to user interaction. Similar behavior is used when editing item metadata at https://github.com/DSpace/DSpace/blob/c0066d2a2fa4afacadba2bf36f20d94cd818217c/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/WorkspaceItemRestRepository.java#L316
As mentioned during the DSpace 7 meeting and described in [DS-4002](https://jira.duraspace.org/browse/DS-4002), the metadata is automatically included in the item and there's no track on the metadata being created automatically. The user just suddenly sees the metadata appear without having neither control nor explanation of the reason why the metadata was created
Can you please explain how this functionality can be altered to give the user the choice on whether the metadata is included in the item?

I don't see any issues for the findByEPerson search method.

**Review by benbosman — COMMENTED** [review]
My impressions of the PR are in line with Tim's review.

After a detailed review of the code in this PR, the code is looking good, but it seems that the functionality in this PR is almost entirely focussed on changes for the automatic enhancement of a submission.
The implementation strategy that was used here (this interpretation is based on the code only due to the lack of comments which were mentioned by @tdonohue ) is:
* PreProcessing functions are created to initialize the listeners for automatically adding the metadata
* PostProcessing functions are created to actually add the automatic metadata

But these PreProcessing and PostProcessing are embedded deeply in the submission backend, and currently don't allow any changes to trigger user interaction. A sample when uploading a file can be seen at https://github.com/DSpace/DSpace/blob/c0066d2a2fa4afacadba2bf36f20d94cd818217c/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/WorkspaceItemRestRepository.java#L244 where the PostProcessing is not supposed to happen prior to user interaction. Similar behavior is used when editing item metadata at https://github.com/DSpace/DSpace/blob/c0066d2a2fa4afacadba2bf36f20d94cd818217c/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/WorkspaceItemRestRepository.java#L316
As mentioned during the DSpace 7 meeting and described in [DS-4002](https://jira.duraspace.org/browse/DS-4002), the metadata is automatically included in the item and there's no track on the metadata being created automatically. The user just suddenly sees the metadata appear without having neither control nor explanation of the reason why the metadata was created
Can you please explain how this functionality can be altered to give the user the choice on whether the metadata is included in the item?

In the WorkspaceItemRestRepository (https://github.com/DSpace/DSpace/blob/c0066d2a2fa4afacadba2bf36f20d94cd818217c/dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/WorkspaceItemRestRepository.java): can you please explain the difference between the two separate upload functions, both calling e.g. the uploadableStep.upload function, but only one using pre and post processing?

I don't see any issues for the findByEPerson search method.

**Comment by abollini** [comment]
I should have addressed all the comments above. Please note that javadoc has been added on the interfaces. Where there was not significant information to provide on the implementation classes only the class javadoc is provided as the interface javadoc is automatically inherited.
@benbosman the two different upload methods are used to bulk create submission or add file to a specific submission, a method was missing in the RestResourceController to make use of them, javadoc provided so it should be more clear now, thanks to note.
A quite large number of ITs have been provided to cover all the proper use of the endpoint, some outstanding issues have been postponed (I hope you agree) to separate tickets see
https://jira.duraspace.org/browse/DS-4042
https://jira.duraspace.org/browse/DS-4043

**Comment by tdonohue** [comment]
@abollini : I haven't done a full re-review yet, but I do see that two classes are still missing requested Javadocs: `ExtractionStep` and `MetadataStep` (which also includes a FIXME near the start of the class that says to "move to ExtractionStep").

Beyond that, glad to see the large number of new Integration Tests (increased 1.5% coverage!!!) and added javadocs.  I'll give this a closer look later today

**Comment by abollini** [comment]
good point @tdonohue sorry to miss that in my previous comment. As you noted the classes missing javadoc are mostly empty. The real classes are the underlying interface AbstractProcessingStep (almost unchanged from the existent code)  and the concrete implementations and new interfaces in the REST maven project org.dspace.app.rest.submit.step.ExtractMetadataStep 
org.dspace.app.rest.submit.step.DescribeStep
org.dspace.app.rest.submit.UploadableStep
these have javadoc but please correct me if I'm wrong.

Please note that the step classes in the dspace-api are now mostly empty and unneeded, this was expected as these classes were put in the dspace-api to provide a common layer between JSPUI and XMLUI but now this common layer (among any future UI) is just the REST API.

This is eventually a subsequent cleanup task, do you agree?

File: dspace-api/src/main/java/org/dspace/submit/step/MetadataStep.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+import org.dspace.core.Utils;
+import org.dspace.services.factory.DSpaceServicesFactory;
+import org.dspace.submit.AbstractProcessingStep;
+import org.dspace.submit.listener.MetadataListener;
+import org.dspace.submit.lookup.SubmissionLookupDataLoader;
+
+/**
+ * @author Luigi Andrea Pascarelli (luigiandrea.pascarelli at 4science.it)
+ */
+public class MetadataStep extends AbstractProcessingStep {
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/submit/step/MetadataStep.java]: @abollini : As this is a non-empty class, I think this one should still have some JavaDocs.  It's not exactly clear what all the roles of the `MetadataStep` are. Currently, it's a 200 line class with no comments.

**Review by tdonohue — APPROVED** [review]
👍 Overall, this code is looking good to me. I added a few new inline comments (on very minor things).  With the addition of all the new ITs and Javadocs though, I think this is looks good enough to merge.  Thanks @abollini 

UPDATE: Also, I agree that the various "nearly empty" step classes (in dspace-api) can be removed in a subsequent cleanup task.  I would recommend though that we create a ticket for that cleanup, unless it is planned to be done very soon -- I don't want it to be forgotten.

**Comment by abollini** [comment]
@tdonohue I have created https://jira.duraspace.org/browse/DS-4045 to remind about the cleanup

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/WorkspaceItemConverter.java [hunk 3]
```diff
@@ -86,6 +85,10 @@ public WorkspaceItemRest fromModel(org.dspace.content.WorkspaceItem obj) {
         // info
 
         if (collection != null) {
+            // we set the status to true as we will discover validation error later in this block
+            // we could eventually leave the status to empty if we don't have collection information, this could be
+            // eventually the case when projection support will be included
```
[the remark this answers is not in the document]
Comment (abollini) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/WorkspaceItemConverter.java]: IMHO it is not a TODO or FIXME right now... it is more a comment to remind my(our)self why it could be useful to have the status as a Boolean instead than a boolean. But if you think more appropriate to flag it as a TODO or FIXME I will make the change

[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/converter/WorkspaceItemConverter.java]: I was just worried that the comment might get lost/forgotten.  Flagging as TODO or FIXME ensures that (at least some) IDEs will highlight the note.  But, I don't feel strongly on this, so I'm fine with leaving this as-is, since you seem to have a plan to clean this up in the near future.

**Comment by tdonohue** [comment]
@benbosman : do you have any objections to merging this?  It looks good to me now, and I believe @abollini has answered all the questions (via the comments above, and the various new tickets linked off of this PR).

**Review by benbosman — APPROVED** [review]
I have no objections to merging this. @abollini has answered the questions.
Thanks for all the updates
